File: src/main/java/org/devcon/ticket/UnpredictableNumberTool.java
package org.devcon.ticket;

import com.alphawallet.attestation.core.AttestationCrypto;
import com.alphawallet.attestation.core.ExceptionUtil;
import com.alphawallet.attestation.core.URLUtility;
import java.nio.ByteBuffer;
import java.nio.charset.StandardCharsets;
import java.security.SecureRandom;
import java.time.Clock;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.bouncycastle.crypto.digests.KeccakDigest;
import org.bouncycastle.crypto.macs.HMac;
import org.bouncycastle.crypto.params.KeyParameter;
import org.tokenscript.eip712.Eip712Common;

public class UnpredictableNumberTool {
  private static final Logger logger = LogManager.getLogger(UnpredictableNumberTool.class);
  public static final int BYTES_IN_SEED = AttestationCrypto.BYTES_IN_DIGEST;
  public static final long DEFAULT_VALIDITY_IN_MS = 3600*1000;
  public static final int BYTES_IN_UN = 8; // 64 bits
  private static final ByteBuffer longBuffer = ByteBuffer.allocate(Long.BYTES);

  private final SecureRandom random;
  private final String domain;
  private final long validityInMs;
  private final HMac hmac = new HMac(new KeccakDigest(256));

  public UnpredictableNumberTool(byte[] key, String domain) {
    this(new SecureRandom(), key, domain);
  }

  public UnpredictableNumberTool(SecureRandom random, byte[] key, String domain) {
    this(random, key, domain, DEFAULT_VALIDITY_IN_MS);
  }

  public UnpredictableNumberTool(SecureRandom random, byte[] key, String domain, long validityInMs) {
    this.random = random;
    this.domain = domain.toLowerCase();
    this.validityInMs = validityInMs;
    hmac.init(new KeyParameter(key));
    // todo should be moved to url utility
    if (!Eip712Common.isDomainValid(domain)) {
      ExceptionUtil.throwException(logger, new IllegalArgumentException("Domain is not a valid domain"));
    }
  }

  public String getDomain() {
    return domain;
  }

  /**
   * Computes an URL friendly base 64 encoding of HMAC(expirationInMs || domain).
   * The encoding consists of BYTES_IN_UN bytes
   * ExpirationInMs is computed from current time plus DEFAULT_VALIDITY_IN_MS
   */
  public UnpredictableNumberBundle getUnpredictableNumberBundle() {
    long expiration = Clock.systemUTC().millis() + validityInMs;
    byte[] randomness = random.generateSeed(BYTES_IN_SEED);
    return new UnpredictableNumberBundle(getUnpredictableNumber(randomness, expiration), randomness, domain, expiration);
  }

  private String getUnpredictableNumber(byte[] randomness, long expirationInMs) {
    hmac.reset();
    hmac.update(longToBytes(expirationInMs), 0, Long.BYTES);
    hmac.update(randomness, 0, BYTES_IN_SEED);
    hmac.update(domain.getBytes(StandardCharsets.UTF_8), 0, domain.getBytes(StandardCharsets.UTF_8).length);
    byte[] digest = new byte[256 / 8];
    hmac.doFinal(digest, 0);
    byte[] result = new byte[BYTES_IN_UN];
    System.arraycopy(digest, 0, result, 0, BYTES_IN_UN);
    return URLUtility.encodeData(result);
  }

  public boolean validateUnpredictableNumber(String un, byte[] randomness, long expirationInMs) {
    if (Clock.systemUTC().millis() > expirationInMs) {
      logger.error("Unpredictable number has expired");
      return false;
    }
    String expectedNumber = getUnpredictableNumber(randomness, expirationInMs);
    if (!expectedNumber.equals(un)) {
      logger.error("The unpredictable number is computed incorrectly. Either wrong key or wrong domain");
      return false;
    }
    return true;
  }

  private static byte[] longToBytes(long x) {
    longBuffer.putLong(0, x);
    return longBuffer.array();
  }
}

File: src/main/javascript/crypto/src/libs/UnpredictableNumberTool.ts
import {Hmac} from 'crypto';
import {bnToUint8, isDomainValid, uint8arrayToBase64} from "./utils";
import {UnpredictableNumberBundle} from "./UnpredictableNumberBundle";
const { TextEncoder } = require('util');

export const DEFAULT_VALIDITY_IN_MS: bigint = BigInt(3600 * 1000);
export const BYTES_IN_UN: number = 8;
export const BYTES_IN_SEED: number = 32;

export class UnpredictableNumberTool {

  private readonly _domain: string;
  private readonly validityInMs: bigint;
  private readonly key: Uint8Array;

  constructor(key: Uint8Array, domain: string, validityInMs: bigint) {
    if (!isDomainValid(domain)) {
      throw Error('Domain is not a valid domain');
    }
    this.key = key;
    this._domain = domain;
    this.validityInMs = validityInMs;
  }

  get domain(): string {
    return this._domain;
  }

  get unpredictableNumberBundle(): UnpredictableNumberBundle {
    const expiration: bigint = BigInt(new Date().getUTCMilliseconds()) + this.validityInMs;
    const randomness = require('secure-random').randomUint8Array(BYTES_IN_SEED);
    let unpredictableNumber = this.getUnpredictableNumber(randomness, expiration);
    return new UnpredictableNumberBundle(unpredictableNumber, randomness, this._domain, expiration);
  }

  private getUnpredictableNumber(randomness: Uint8Array, expirationInMs: bigint): string {
    const hmac: Hmac = require('crypto').createHmac('sha3-256', this.key);
    hmac.update(bnToUint8(expirationInMs));
    hmac.update(randomness);
    hmac.update(new TextEncoder().encode(this._domain));
    const digest: Buffer = hmac.digest();
    const result = digest.slice(0, BYTES_IN_UN);
    return uint8arrayToBase64(result);
  }

  validateUnpredictableNumber(un: string, randomness: Uint8Array, expirationInMs: bigint): boolean {
    if (new Date().getUTCMilliseconds() > expirationInMs) {
      console.error('Unpredictable number has expired');
      return false;
    }
    const expectedNumber = this.getUnpredictableNumber(randomness, expirationInMs);
    if (expectedNumber !== un) {
      console.error('The unpredictable number is computed incorrectly. Either wrong key or wrong domain');
      return false;
    }
    return true;
  }
}
